### ISSUE TITLE
OptimizerBase.CalculateGradient fallback uses algebraically wrong formula + latent over-indexing bug

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1837]
## Problem

The virtual fallback `CalculateGradient(solution, xTrain, yTrain, batchIndices)` at `src/Optimizers/GradientBasedOptimizerBase.cs:1491` uses an **algebraically wrong** gradient formula and has a **latent over-indexing bug** on top of it.

### The wrong formula

```csharp
// src/Optimizers/GradientBasedOptimizerBase.cs:1512-1531
for (int i = 0; i < batchIndices.Length; i++)
{
    var input = InputHelper<T, TInput>.GetItem(xBatch, i);
    var target = InputHelper<T, TOutput>.GetItem(yBatch, i);
    var prediction = solution.Predict(input);
    var error = LossFunction.CalculateLoss(
        ConversionsHelper.ConvertToVector<T, TOutput>(prediction),
        ConversionsHelper.ConvertToVector<T, TOutput>(target));

    // Update gradient based on the error
    for (int j = 0; j < parameters.Length; j++)
    {
        var featureValue = InputHelper<T, TInput>.GetFeatureValue(input, j);
        var contribution = NumOps.Multiply(error, featureValue);
        gradient[j] = NumOps.Add(gradient[j], contribution);
    }
}
```

Two overlapping bugs:

1. **`error` is `LossFunction.CalculateLoss(...)` — a scalar loss.** For MSE, that's the mean of squared errors, NOT the per-element error signal `(pred - target)`. Multiplying a scalar loss by an input feature and calling it a gradient is bogus.
2. **Even the intended formula is wrong for neural networks.** `gradient[j] = error × input_feature[j]` is the linear-regression gradient formula (and only correct if `error` were the per-sample residual, which it isn't). For any multi-layer model, gradient requires backprop through the chain rule. This isn't a "close enough" approximation — it's the wrong function class.
3. **Over-indexing.** The inner loop runs `j` from `0` to `parameters.Length`. For a NeRF with 348k trainable parameters but a 6-dimensional input, `InputHelper<T, TInput>.GetFeatureValue(input, j)` for `j >= input_dim` either throws or returns garbage. Silently bakes into the "gradient" if it doesn't throw.

### Why it hasn't blown up yet

The 3-arg overload at line 960 (`CalculateGradient(solution, X, y)`) takes the good path via `IGradientComputable`:

```csharp
if (solution is IGradientComputable<T, TInput, TOutput> gradientComputable)
{
    gradient = gradientComputable.ComputeGradients(X, y, LossFunction);
    // ...
}
```

`NeuralNetworkBase<T>` implements `IGradientComputable<T, Tensor<T>, Tensor<T>>` (see the top of `src/NeuralNetworks/NeuralNetworkBase.cs`), so every neural-network model takes the correct path. The 4-arg virtual is a fallback that's currently unreachable in the shipped model set — but it's a **landmine**: any subclass that overrides this virtual "to extend the base," or any future model class that skips `IGradientComputable`, silently gets algebraic garbage for gradients.

## Fix direction (agreed in design discussion)

**Delete the wrong-formula body. Replace with a fail-fast throw routed through `InterfaceGuard`.**

Rationale:

- **Industry-standard behavior for "no gradient method available"**: PyTorch, Keras, JAX all throw a clear error. AiDotNet should match; there's no silent-fallback pattern to preserve.
- **The current fallback is unreachable in the shipped set** — nothing to migrate. Deleting is safe.
- **`InterfaceGuard` already provides the pattern** — `InterfaceGuard.Parameterizable(...)` and other guard methods emit uniform messages across the codebase. Add `InterfaceGuard.RequireGradientComputable<T, TInput, TOutput>(solution)` for symmetry.

Concrete replacement:

```csharp
protected virtual Vector<T> CalculateGradient(
    IFullModel<T, TInput, TOutput> solution,
    TInput xTrain,
    TOutput yTrain,
    int[] batchIndices)
{
    throw new NotSupportedException(
        $"Model {solution.GetType().Name} does not implement " +
        $"IGradientComputable<{typeof(T).Name}, {typeof(TInput).Name}, {typeof(TOutput).Name}>. " +
        $"Gradient-based optimizers require a model that can compute gradients via backprop. " +
        $"Inherit from NeuralNetworkBase<T> (which implements IGradientComputable) or " +
        $"implement IGradientComputable<T, TInput, TOutput> directly.");
}
```

## Beyond-industry notes

The mainstream frameworks (PyTorch, Keras, JAX) don't offer a first-class finite-difference gradient fallback either. That WOULD be a beyond-industry feature — an opt-in `AllowFiniteDifferenceGradientFallback = true` on `AdamOptimizerOptions` that lets scientific-ML users with non-differentiable models (physics simulators, black-box optimizers) train via numerical gradients at 10-100× slowdown but with a clear diagnostic warning.

**Explicitly out of scope for this bug fix** — noting as a future feature so it's tracked. If a legitimate use case emerges, file separately as an enhancement.

## Wiring surface

No user-facing surface changes. Purely internal: delete the wrong body, throw. Callers using `IGradientComputable`-implementing models see zero difference.

## Related

- #1826 / #1829 — original facade unblock chain (this was surfaced during the deep dive into `OptimizerBase` for the feature-selection fix)
- No user-facing impact today because the code path is unreachable for shipped models, but the landmine risk grows every time a new model class ships in AiDotNet.

**Comment by ooples** [comment]
**Already fixed — but on `master`, not by PR #1789.** The `CalculateGradient(solution, xTrain, yTrain, batchIndices)` fallback now routes through `InterfaceGuard.GradientComputable(solution).ComputeGradients(xBatch, yBatch, LossFunction)`. I checked `master` directly: the per-sample loop that used `LossFunction.CalculateLoss(...)` as the *error* term, and the `GetItem(xBatch, i)` over-indexing alongside it, are both gone there. Deliberately **not** linked to #1789 because that PR is not what fixed it — this can be closed independently.

**Comment by ooples** [comment]
Closing: verified fixed on `master`. The `CalculateGradient(solution, xTrain, yTrain, batchIndices)` fallback now routes through `InterfaceGuard.GradientComputable(solution).ComputeGradients(xBatch, yBatch, LossFunction)`. Checked `master` directly — both defects reported here are gone: the per-sample loop that used `LossFunction.CalculateLoss(...)` as the *error* term, and the `GetItem(xBatch, i)` over-indexing alongside it. Not attributable to PR #1789, hence closed independently rather than linked to it.
